### PR TITLE
Exception logging for MessageReceiver background exceptions

### PR DESCRIPTION
**Pull request by mathewc**
Merging https://github.com/Azure/azure-webjobs-sdk/pull/1704 to dev. Code is a bit different in dev branch due to API differences.

### CODE REVIEW

File: sample/SampleHost/Functions.cs [hunk 1]
```diff
@@ -33,5 +35,18 @@ public static class Functions
...
+
+        public static async Task ProcessWorkItem_ServiceBus(
+            [ServiceBusTrigger("test-items")] WorkItem item,
+            string messageId,
+            int deliveryCount,
+            ILogger log)
+        {
+            log.LogInformation($"Processing ServiceBus message (Id={messageId}, DeliveryCount={deliveryCount})");
+
+            await Task.Delay(TimeSpan.FromMinutes(7));
```
Comment (mathewc) [inline sample/SampleHost/Functions.cs]: Will revert this back to 1 second after checkin. Still testing lock loss scenarios + logging

File: src/Microsoft.Azure.WebJobs.Host/Loggers/Logger/Constants/LogCategories.cs [hunk 2]
```diff
@@ -24,6 +24,11 @@ public static class LogCategories
         /// </summary>
         public const string Singleton = "Host.Singleton";
 
+        /// <summary>
+        /// The category for all logs written by the function executor.
+        /// </summary>
+        public const string Executor = "Host.Executor";
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Loggers/Logger/Constants/LogCategories.cs]: This exists in the v2 branch, but not here for some reason

File: src/Microsoft.Azure.WebJobs.ServiceBus/Config/ServiceBusExtensionConfig.cs [hunk 3]
```diff
@@ -54,6 +57,20 @@ public void Initialize(ExtensionConfigContext context)
                 throw new ArgumentNullException("context");
             }
 
+            // Register an exception handler for background exceptions
+            // coming from MessageReceivers.
+            Config.ExceptionHandler = (e) =>
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/Config/ServiceBusExtensionConfig.cs]: Our setting this here will override the logic used by our default MessageReceiverOptions that we set up. If the user sets ServiceBusConfiguration.MessageOptions to their own instance with their own handler, they completely override, and our handler doesn't get used (by design). For WebJobs scenarios that's what we want - user has full control. For Functions, they can't override our MessageReceiverOptions instance, so our logging will always be in effect.

File: src/Microsoft.Azure.WebJobs.ServiceBus/Config/ServiceBusConfiguration.cs [hunk 4]
```diff
@@ -54,7 +53,7 @@ public string ConnectionString
         }
 
         /// <summary>
-        /// Gets or sets the default <see cref="MessageHandlerOptions"/> that will be used by
+        /// Gets the default <see cref="MessageHandlerOptions"/> that will be used by
```
Comment (fabiocav) [inline src/Microsoft.Azure.WebJobs.ServiceBus/Config/ServiceBusConfiguration.cs]: nit: This still has a setter.

File: src/Microsoft.Azure.WebJobs.ServiceBus/Config/ServiceBusExtensionConfig.cs [hunk 5]
```diff
@@ -54,6 +57,20 @@ public void Initialize(ExtensionConfigContext context)
...
+            {
+                var sbex = e.Exception as ServiceBusException;
```
Comment (fabiocav) [inline src/Microsoft.Azure.WebJobs.ServiceBus/Config/ServiceBusExtensionConfig.cs]: nit: you should be able to inline this (e.g. `(if e.Exception is ServiceBusException sbx && !sbex.IsTransient)`)

Comment (mathewc) [reply src/Microsoft.Azure.WebJobs.ServiceBus/Config/ServiceBusExtensionConfig.cs]: But then I have to cast the exception twice

File: src/Microsoft.Azure.WebJobs.ServiceBus/Config/ServiceBusConfiguration.cs (hunk 4 above)
Comment (mathewc) [reply src/Microsoft.Azure.WebJobs.ServiceBus/Config/ServiceBusConfiguration.cs]: Fixed

**Pull request merged**